src/Recoil__State.re

```diff
@@ -0,0 +1,19 @@
+type t = string;
+type view =
+  | Loading
+  | HasValue
+  | HasError;
+
+[@bs.inline]
+let loading = "loading";
+[@bs.inline]
+let hasValue = "hasValue";
+[@bs.inline]
+let hasError = "hasError";
+
+let view =
+  fun
+  | "loading" => Loading
+  | "hasValue" => HasValue
+  | "hasError" => HasError
+  | other => failwith("unrecognized State value: " ++ other);
```

Comment:
This introduces a bit of runtime but allows to pattern match:

```js
function view(other) {
  switch (other) {
    case "hasError" :
        return /* HasError */2;
    case "hasValue" :
        return /* HasValue */1;
    case "loading" :
        return /* Loading */0;
    default:
      return Pervasives.failwith("unrecognized State value: " + other);
  }
}
```

If you prefer I can make them just inline string and one can do:

```js
if(loadableUser.state == Recoil.State.hasValue){...}
```

Reply:
Yeah, I'd go with the runtime-free solution. I initially didn't implement it because you can pretty much do the same with `Suspense` and error boundaries, which (I think) will be most of the usage.